File: src/main/java/no/nav/fo/veilarbregistrering/bruker/Identer.java
package no.nav.fo.veilarbregistrering.bruker;

import javax.ws.rs.NotFoundException;
import java.util.List;
import java.util.stream.Collectors;

public class Identer {
    private List<Ident> identer;

    public Identer(List<Ident> identer) {
        this.identer = identer;
    }

    public static Identer of(List<Ident> identer) {
        return new Identer(identer);
    }

    public List<Ident> getIdenter() {
        return identer;
    }

    public Foedselsnummer finnGjeldendeFnr() {
        Ident gjeldendeFnr = identer.stream()
                .filter(ident -> ident.getGruppe() == Gruppe.FOLKEREGISTERIDENT)
                .filter(ident -> !ident.isHistorisk())
                .findFirst()
                .orElseThrow(() -> new NotFoundException("Kunne ikke finne et gjeldende fødselsnummer"));
        return Foedselsnummer.of(gjeldendeFnr.getIdent());
    }

    public AktorId finnGjeldendeAktorId() {
        Ident gjeldendeAktorId = identer.stream()
                .filter(ident -> ident.getGruppe() == Gruppe.AKTORID)
                .filter(ident -> !ident.isHistorisk())
                .findFirst()
                .orElseThrow(() -> new NotFoundException("Kunne ikke finne en gjeldende aktørId"));
        return AktorId.of(gjeldendeAktorId.getIdent());
    }

    public List<Foedselsnummer> finnHistoriskeFoedselsnummer() {
        return identer.stream()
                .filter(ident -> ident.getGruppe() == Gruppe.FOLKEREGISTERIDENT)
                .filter(Ident::isHistorisk)
                .map(Ident::getIdent)
                .map(Foedselsnummer::of)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/bruker/Identer.kt
package no.nav.fo.veilarbregistrering.bruker

import no.nav.fo.veilarbregistrering.bruker.feil.ManglendeBrukerInfoException

class Identer(val identer: List<Ident>) {
    fun finnGjeldendeFnr(): Foedselsnummer {
        val gjeldendeFnr = identer
            .filter { it.gruppe == Gruppe.FOLKEREGISTERIDENT }
            .firstOrNull { !it.isHistorisk }
            ?: throw ManglendeBrukerInfoException("Kunne ikke finne et gjeldende fødselsnummer")
        return Foedselsnummer(gjeldendeFnr.ident)
    }

    fun finnGjeldendeAktorId(): AktorId {
        val gjeldendeAktorId = identer
            .filter { it.gruppe == Gruppe.AKTORID }
            .firstOrNull { !it.isHistorisk }
            ?: throw ManglendeBrukerInfoException("Kunne ikke finne en gjeldende aktørId")
        return AktorId(gjeldendeAktorId.ident)
    }

    fun finnHistoriskeFoedselsnummer(): List<Foedselsnummer> =
        identer
            .filter { it.gruppe == Gruppe.FOLKEREGISTERIDENT && it.isHistorisk}
            .map { it.ident }
            .map { Foedselsnummer(it) }
}

class Ident(val ident: String, val isHistorisk: Boolean, val gruppe: Gruppe)

enum class Gruppe {
    FOLKEREGISTERIDENT, AKTORID, NPID
}